TripalCultivate/TripalCultivate-Phenotypes
# G2.39 template generator

**Issue #39** Data Collection Template File Generator

## Motivation



Create a data collection template file in tab-separated format (tsv). The headers encoded in the template file are the headers defined by the headers property in the importer instance.

Template file can be downloaded by clicking on the the link Download a template file located just below the list of headers of the importer.

## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

1. Resolve the value for the href attribute of the Download template file link in the header listing of the importer.
![image](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/15472253/96b2cab3-18af-41e9-90f2-78456613019d)
2. The link is the generated template file by a service.
3. Headers defined in the Importer through the headers property are then encoded into the file as the headers column. Headers are converted to required format of tab-separated format. 

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

1. Functional tests included as part of the Importer API functional test and check if the template file was created.
2. Kernel tests test the template generator service to create a file.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Switch to branch g2.39-Template-Generator.  Install and enable the module
3. Access Tripal page that displays all implementation of Importer plugin in admin/tripal/loaders
4. Click on Tripal Cultivate: Open Science Phenotypic Data link
5. In stage 1, under the list of required column headers, click the link Download a template file link.
